src/jit/optimizer.cpp

```diff
@@ -1416,410 +1418,1018 @@ void Compiler::optCheckPreds()
 
 #endif // DEBUG
 
-/*****************************************************************************
- * Find the natural loops, using dominators. Note that the test for
- * a loop is slightly different from the standard one, because we have
- * not done a depth first reordering of the basic blocks.
- */
-
-void Compiler::optFindNaturalLoops()
+namespace
 {
-#ifdef DEBUG
-    if (verbose)
+//------------------------------------------------------------------------
+// LoopSearch: Class that handles scanning a range of blocks to detect a loop,
+//             moving blocks to make the loop body contiguous, and recording
+//             the loop.
+//
+// We will use the following terminology:
+//   HEAD    - the basic block that flows into the loop ENTRY block (Currently MUST be lexically before entry).
+//             Not part of the looping of the loop.
+//   FIRST   - the lexically first basic block (in bbNext order) within this loop.  (May be part of a nested loop,
+//             but not the outer loop. ???)
+//   TOP     - the target of the backward edge from BOTTOM. In most cases FIRST and TOP are the same.
+//   BOTTOM  - the lexically last block in the loop (i.e. the block from which we jump to the top)
+//   EXIT    - the predecessor of loop's unique exit edge, if it has a unique exit edge; else nullptr
+//   ENTRY   - the entry in the loop (not necessarly the TOP), but there must be only one entry
+//
+//   We (currently) require the body of a loop to be a contiguous (in bbNext order) sequence of basic blocks.
+//   When the loop is identified, blocks will be moved out to make it a compact contiguous region if possible,
+//   and in cases where compaction is not possible, we'll subsequently treat all blocks in the lexical range
+//   between TOP and BOTTOM as part of the loop even if they aren't part of the SCC.
+//
+//        |
+//        v
+//      head
+//        |
+//        |    top/beg <--+
```

Comment:
Fixed in https://github.com/dotnet/coreclr/pull/13314/commits/73f116794e24aa5517dca28914ce2a7889634539